pkg/virt-operator/creation/components/secrets.go

```diff
@@ -0,0 +1,311 @@
+	}
+	crt, certKey, ca := strategy(secret, caCert, duration.Duration)
+
+	secret.Data = map[string][]byte{
+		bootstrap.CertBytesValue:        cert.EncodeCertPEM(crt),
+		bootstrap.KeyBytesValue:         cert.EncodePrivateKeyPEM(certKey),
+		bootstrap.SigningCertBytesValue: cert.EncodeCertPEM(ca),
+	}
+
+	if secret.Annotations == nil {
+		secret.Annotations = map[string]string{}
+	}
+	secret.Annotations["kubevirt.io/duration"] = duration.String()
+	return nil
+}
+
+func NewCACertSecret(operatorNamespace string) *k8sv1.Secret {
+	return &k8sv1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      KubeVirtCASecretName,
+			Namespace: operatorNamespace,
+			Labels: map[string]string{
+				v1.ManagedByLabel: v1.ManagedByLabelOperatorValue,
+			},
+		},
+		Type: "Opaque",
+	}
+}
+
+func NewCertSecrets(installNamespace string, operatorNamespace string) []*k8sv1.Secret {
+	secrets := []*k8sv1.Secret{
+
+		{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Secret",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      VirtApiCertSecretName,
+				Namespace: installNamespace,
+				Labels: map[string]string{
+					v1.ManagedByLabel: v1.ManagedByLabelOperatorValue,
+				},
+			},
+			Type: "Opaque",
+		},
+		{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Secret",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      VirtControllerCertSecretName,
+				Namespace: installNamespace,
+				Labels: map[string]string{
+					v1.ManagedByLabel: v1.ManagedByLabelOperatorValue,
+				},
+			},
+			Type: "Opaque",
+		},
+		{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Secret",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      VirtHandlerServerCertSecretName,
+				Namespace: installNamespace,
+				Labels: map[string]string{
+					v1.ManagedByLabel: v1.ManagedByLabelOperatorValue,
+				},
+			},
+			Type: "Opaque",
+		},
+		{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Secret",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      VirtHandlerCertSecretName,
+				Namespace: installNamespace,
+				Labels: map[string]string{
+					v1.ManagedByLabel: v1.ManagedByLabelOperatorValue,
+				},
+			},
+			Type: "Opaque",
+		},
+		{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "Secret",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      VirtOperatorCertSecretName,
+				Namespace: operatorNamespace,
+				Labels: map[string]string{
+					v1.ManagedByLabel: v1.ManagedByLabelOperatorValue,
+				},
+			},
+			Type: "Opaque",
+		},
+	}
+	return secrets
+}
+
+// nextRotationDeadline returns a value for the threshold at which the
+// current certificate should be rotated, 80% of the expiration of the
+// certificate.
+func NextRotationDeadline(cert *tls.Certificate, ca *tls.Certificate) time.Time {
```

Comment:
The library CDI uses stores the cert NotBefore/NotAfter in annotations.  What's nice about it is that rotation can be forced if necessary by simply updating the annotation.  This was how I demo'd cert rotation.

Reply:
I have seen that. I am undecided on this. The certificates could be messed up and no one may notice it because the annotations say that it is still valid. If I then in addition check the certificate anyway, I am not sure I need the annotations. For demoing I can just set the rotation configs to very short intervals.